storage/block/wired_list.go

```diff
@@ -0,0 +1,269 @@
+	evicted              tally.Counter
+	pushedBack           tally.Counter
+	inserted             tally.Counter
+	evictedAfterDuration tally.Timer
+}
+
+// NewWiredList returns a new database block wired list.
+func NewWiredList(
+	runtimeOptsMgr runtime.OptionsManager,
+	iopts instrument.Options,
+	copts clock.Options,
+) *WiredList {
+	scope := iopts.MetricsScope().
+		SubScope("wired-list")
+	l := &WiredList{
+		nowFn: copts.NowFn(),
+		metrics: wiredListMetrics{
+			// Keeps track of how many blocks are in the list
+			unwireable: scope.Gauge("unwireable"),
+			limit:      scope.Gauge("limit"),
+			// Incremented when a block is evicted
+			evicted: scope.Counter("evicted"),
+			// Incremented when a block is "pushed back" in the list, I.E
+			// it was already in the list
+			pushedBack: scope.Counter("pushed-back"),
+			// Incremented when a block is inserted into the list, I.E
+			// it wasn't already present
+			inserted: scope.Counter("inserted"),
+			// Measure how much time blocks spend in the list before being evicted
+			evictedAfterDuration: scope.Timer("evicted-after-duration"),
+		},
+	}
+	l.root.setNext(&l.root)
+	l.root.setPrev(&l.root)
+	runtimeOptsMgr.RegisterListener(l)
+	return l
+}
+
+// SetRuntimeOptions sets the current runtime options to
+// be consumed by the wired list
+func (l *WiredList) SetRuntimeOptions(value runtime.Options) {
+	atomic.StoreInt64(&l.maxWired, int64(value.MaxWiredBlocks()))
+}
+
+// Start starts processing the wired list
+func (l *WiredList) Start() {
+	l.Lock()
+	defer l.Unlock()
+	if l.updatesCh != nil {
+		return
+	}
+
+	l.updatesCh = make(chan DatabaseBlock, wiredListEventsChannelLength)
+	l.doneCh = make(chan struct{}, 1)
+	go func() {
+		i := 0
+		for v := range l.updatesCh {
+			l.processUpdateBlock(v)
+			if i%wiredListSampleGaugesEvery == 0 {
+				l.metrics.unwireable.Update(float64(l.length))
+				l.metrics.limit.Update(float64(atomic.LoadInt64(&l.maxWired)))
+			}
+			i++
+		}
+		l.doneCh <- struct{}{}
+	}()
+}
+
+// Stop stops processing the wired list
+func (l *WiredList) Stop() {
+	close(l.updatesCh)
+	<-l.doneCh
+
+	l.Lock()
+	defer l.Unlock()
+	l.updatesCh = nil
+	close(l.doneCh)
+	l.doneCh = nil
+}
+
+// Update places the block into the channel of blocks which are waiting to notify the
+// wired list that they were accessed. All updates must be processed through this channel
+// to force synchronization.
+//
+// We use a channel and a background processing goroutine to reduce blocking / lock contention.
+func (l *WiredList) Update(v DatabaseBlock) {
+	l.updatesCh <- v
+}
+
+// processUpdateBlock inspects a block that has been modified or read recently
+// and determines what outcome its state should have on the wired list.
+func (l *WiredList) processUpdateBlock(v DatabaseBlock) {
+	entry := v.wiredListEntry()
+	// The WiredList should never receive closed blocks or blocks that were not retrieved
+	// from disk, but we include the sanity check for posterity.
```

Comment:
If that's so, we should probably emit a metric and/or log a warning when this happens.

Reply:
done